File: ios/HuyaMonitor/Tars.swift
import Foundation

enum Tars {
    static let byte = 0
    static let short = 1
    static let int = 2
    static let long = 3
    static let float = 4
    static let double = 5
    static let string1 = 6
    static let string4 = 7
    static let map = 8
    static let list = 9
    static let structBegin = 10
    static let structEnd = 11
    static let zero = 12
    static let simpleList = 13
}

final class TarsOutput {
    private(set) var bytes = [UInt8]()

    func toData() -> Data {
        return Data(bytes)
    }

    private func appendBigEndian(_ value: UInt64, count: Int) {
        var shift = (count - 1) * 8
        while shift >= 0 {
            bytes.append(UInt8((value >> UInt64(shift)) & 0xFF))
            shift -= 8
        }
    }

    func writeHead(_ type: Int, tag: Int) {
        if tag < 15 {
            bytes.append(UInt8((tag << 4) | type))
        } else {
            bytes.append(UInt8(0xF0 | type))
            bytes.append(UInt8(tag & 0xFF))
        }
    }

    func writeBool(_ value: Bool, tag: Int) {
        writeInt(value ? 1 : 0, tag: tag)
    }

    func writeInt(_ value: Int, tag: Int) {
        if value == 0 {
            writeHead(Tars.zero, tag: tag)
            return
        }
        if value >= -128 && value <= 127 {
            writeHead(Tars.byte, tag: tag)
            bytes.append(UInt8(truncatingIfNeeded: value))
            return
        }
        if value >= -32768 && value <= 32767 {
            writeHead(Tars.short, tag: tag)
            appendBigEndian(UInt64(UInt16(truncatingIfNeeded: value)), count: 2)
            return
        }
        if value >= -2147483648 && value <= 2147483647 {
            writeHead(Tars.int, tag: tag)
            appendBigEndian(UInt64(UInt32(truncatingIfNeeded: value)), count: 4)
            return
        }
        writeHead(Tars.long, tag: tag)
        appendBigEndian(UInt64(bitPattern: Int64(value)), count: 8)
    }

    func writeString(_ text: String, tag: Int) {
        let data = Array(text.utf8)
        if data.count > 255 {
            writeHead(Tars.string4, tag: tag)
            appendBigEndian(UInt64(UInt32(data.count)), count: 4)
        } else {
            writeHead(Tars.string1, tag: tag)
            bytes.append(UInt8(data.count))
        }
        bytes.append(contentsOf: data)
    }

    func writeBytes(_ data: Data, tag: Int) {
        let payload = [UInt8](data)
        writeHead(Tars.simpleList, tag: tag)
        writeHead(Tars.byte, tag: 0)
        writeInt(payload.count, tag: 0)
        bytes.append(contentsOf: payload)
    }
}

final class TarsInput {
    private let data: [UInt8]
    private var pos = 0

    init(_ data: Data) {
        self.data = [UInt8](data)
    }

    init(_ data: [UInt8]) {
        self.data = data
    }

    func remaining() -> Int {
        return data.count - pos
    }

    private func readU8() throws -> UInt8 {
        guard pos < data.count else { throw TarsError.overflow }
        let value = data[pos]
        pos += 1
        return value
    }

    private func readBytes(_ count: Int) throws -> [UInt8] {
        guard count >= 0, pos + count <= data.count else { throw TarsError.overflow }
        let chunk = Array(data[pos..<(pos + count)])
        pos += count
        return chunk
    }

    func readHead() throws -> (type: Int, tag: Int) {
        let first = try readU8()
        let type = Int(first & 0x0F)
        var tag = Int((first & 0xF0) >> 4)
        if tag == 15 {
            tag = Int(try readU8())
        }
        return (type, tag)
    }

    func peekHead() -> (type: Int, tag: Int)? {
        let saved = pos
        defer { pos = saved }
        return try? readHead()
    }

    func skipField(_ type: Int) throws {
        switch type {
        case Tars.byte:
            _ = try readU8()
        case Tars.short:
            _ = try readBytes(2)
        case Tars.int, Tars.float:
            _ = try readBytes(4)
        case Tars.long, Tars.double:
            _ = try readBytes(8)
        case Tars.string1:
            let count = Int(try readU8())
            _ = try readBytes(count)
        case Tars.string4:
            let head = try readBytes(4)
            let count = Int(UInt32(head[0]) << 24 | UInt32(head[1]) << 16 | UInt32(head[2]) << 8 | UInt32(head[3]))
            _ = try readBytes(count)
        case Tars.map:
            let size = readInt(tag: 0)
            for _ in 0..<max(0, size) {
                let keyHead = try readHead()
                try skipField(keyHead.type)
                let valueHead = try readHead()
                try skipField(valueHead.type)
            }
        case Tars.list:
            let size = readInt(tag: 0)
            for _ in 0..<max(0, size) {
                let head = try readHead()
                try skipField(head.type)
            }
        case Tars.structBegin:
            while true {
                let head = try readHead()
                if head.type == Tars.structEnd { break }
                try skipField(head.type)
            }
        case Tars.structEnd, Tars.zero:
            return
        case Tars.simpleList:
            _ = try readHead()
            let count = readInt(tag: 0)
            _ = try readBytes(count)
        default:
            throw TarsError.unknownType(type)
        }
    }

    @discardableResult
    func skipToTag(_ tag: Int) throws -> Bool {
        while remaining() > 0 {
            guard let head = peekHead() else { return false }
            if head.type == Tars.structEnd { return false }
            if head.tag == tag { return true }
            if head.tag > tag { return false }
            _ = try readHead()
            try skipField(head.type)
        }
        return false
    }

    func readInt(tag: Int, fallback: Int = 0) -> Int {
        do {
            guard try skipToTag(tag) else { return fallback }
            let head = try readHead()
            switch head.type {
            case Tars.zero:
                return 0
            case Tars.byte:
                let value = Int(try readU8())
                return value > 127 ? value - 256 : value
            case Tars.short:
                let raw = try readBytes(2)
                let value = UInt16(raw[0]) << 8 | UInt16(raw[1])
                return Int(Int16(bitPattern: value))
            case Tars.int:
                let raw = try readBytes(4)
                let value = UInt32(raw[0]) << 24 | UInt32(raw[1]) << 16 | UInt32(raw[2]) << 8 | UInt32(raw[3])
                return Int(Int32(bitPattern: value))
            case Tars.long:
                var value: UInt64 = 0
                for byte in try readBytes(8) {
                    value = (value << 8) | UInt64(byte)
                }
                return Int(Int64(bitPattern: value))
            default:
                return fallback
            }
        } catch {
            return fallback
        }
    }

    func readBool(tag: Int, fallback: Bool = false) -> Bool {
        return readInt(tag: tag, fallback: fallback ? 1 : 0) != 0
    }

    func readString(tag: Int, fallback: String = "") -> String {
        do {
            guard try skipToTag(tag) else { return fallback }
            let head = try readHead()
            switch head.type {
            case Tars.zero:
                return ""
            case Tars.string1:
                let count = Int(try readU8())
                return String(decoding: try readBytes(count), as: UTF8.self)
            case Tars.string4:
                let raw = try readBytes(4)
                let count = Int(UInt32(raw[0]) << 24 | UInt32(raw[1]) << 16 | UInt32(raw[2]) << 8 | UInt32(raw[3]))
                return String(decoding: try readBytes(count), as: UTF8.self)
            default:
                return fallback
            }
        } catch {
            return fallback
        }
    }

    func readBytes(tag: Int) -> Data {
        do {
            guard try skipToTag(tag) else { return Data() }
            let head = try readHead()
            switch head.type {
            case Tars.zero:
                return Data()
            case Tars.simpleList:
                _ = try readHead()
                let count = readInt(tag: 0)
                return Data(try readBytes(count))
            case Tars.list:
                let count = readInt(tag: 0)
                var out = [UInt8]()
                out.reserveCapacity(max(0, count))
                for _ in 0..<max(0, count) {
                    out.append(UInt8(truncatingIfNeeded: readInt(tag: 0)))
                }
                return Data(out)
            default:
                return Data()
            }
        } catch {
            return Data()
        }
    }

    func readStructBytes(tag: Int) -> Data? {
        do {
            guard try skipToTag(tag) else { return nil }
            let head = try readHead()
            guard head.type == Tars.structBegin else { return nil }
            let start = pos
            var depth = 1
            while depth > 0 {
                let nested = try readHead()
                if nested.type == Tars.structBegin {
                    depth += 1
                } else if nested.type == Tars.structEnd {
                    depth -= 1
                } else {
                    try skipField(nested.type)
                }
            }
            guard pos - 1 >= start else { return nil }
            return Data(data[start..<(pos - 1)])
        } catch {
            return nil
        }
    }
}

enum TarsError: Error {
    case overflow
    case unknownType(Int)
}

File: tars.py
import struct

BYTE = 0
SHORT = 1
INT = 2
LONG = 3
FLOAT = 4
DOUBLE = 5
STRING1 = 6
STRING4 = 7
MAP = 8
LIST = 9
STRUCT_BEGIN = 10
STRUCT_END = 11
ZERO = 12
SIMPLE_LIST = 13


class TarsOutput:
    def __init__(self):
        self.buf = bytearray()

    def write_head(self, ttype, tag):
        if tag < 15:
            self.buf.append((tag << 4) | ttype)
        else:
            self.buf.append(0xF0 | ttype)
            self.buf.append(tag & 0xFF)

    def write_bool(self, value, tag):
        self.write_int(1 if value else 0, tag)

    def write_int(self, n, tag):
        n = int(n)
        if n == 0:
            self.write_head(ZERO, tag)
            return
        if -128 <= n <= 127:
            self.write_head(BYTE, tag)
            self.buf.append(n & 0xFF)
            return
        if -32768 <= n <= 32767:
            self.write_head(SHORT, tag)
            self.buf += struct.pack(">h", n)
            return
        if -2147483648 <= n <= 2147483647:
            self.write_head(INT, tag)
            self.buf += struct.pack(">i", n)
            return
        self.write_head(LONG, tag)
        self.buf += struct.pack(">q", n)

    def write_string(self, text, tag):
        data = text.encode("utf-8")
        if len(data) > 255:
            self.write_head(STRING4, tag)
            self.buf += struct.pack(">I", len(data))
        else:
            self.write_head(STRING1, tag)
            self.buf.append(len(data))
        self.buf += data

    def write_bytes(self, data, tag):
        self.write_head(SIMPLE_LIST, tag)
        self.write_head(BYTE, 0)
        self.write_int(len(data), 0)
        self.buf += data

    def write_struct_begin(self, tag):
        self.write_head(STRUCT_BEGIN, tag)

    def write_struct_end(self):
        self.write_head(STRUCT_END, 0)

    def to_bytes(self):
        return bytes(self.buf)


class TarsInput:
    def __init__(self, data):
        self.data = data if isinstance(data, (bytes, bytearray)) else bytes(data)
        self.pos = 0

    def remaining(self):
        return len(self.data) - self.pos

    def _u8(self):
        if self.pos >= len(self.data):
            raise EOFError("tars buffer overflow")
        b = self.data[self.pos]
        self.pos += 1
        return b

    def _read(self, n):
        if self.pos + n > len(self.data):
            raise EOFError("tars buffer overflow")
        chunk = self.data[self.pos:self.pos + n]
        self.pos += n
        return chunk

    def peek_head(self):
        saved = self.pos
        try:
            return self.read_head()
        finally:
            self.pos = saved

    def read_head(self):
        b = self._u8()
        ttype = b & 0x0F
        tag = (b & 0xF0) >> 4
        if tag == 15:
            tag = self._u8()
        return ttype, tag

    def skip_field(self, ttype=None):
        if ttype is None:
            ttype, _ = self.read_head()
        if ttype == BYTE:
            self._u8()
        elif ttype == SHORT:
            self._read(2)
        elif ttype == INT:
            self._read(4)
        elif ttype == LONG:
            self._read(8)
        elif ttype == FLOAT:
            self._read(4)
        elif ttype == DOUBLE:
            self._read(8)
        elif ttype == STRING1:
            n = self._u8()
            self._read(n)
        elif ttype == STRING4:
            n = struct.unpack(">I", self._read(4))[0]
            self._read(n)
        elif ttype == MAP:
            size = self.read_int(0, False)
            for _ in range(size):
                self.skip_field()
                self.skip_field()
        elif ttype == LIST:
            size = self.read_int(0, False)
            for _ in range(size):
                self.skip_field()
        elif ttype == STRUCT_BEGIN:
            while True:
                nested, _ = self.read_head()
                if nested == STRUCT_END:
                    break
                self.skip_field(nested)
        elif ttype == STRUCT_END:
            return
        elif ttype == ZERO:
            return
        elif ttype == SIMPLE_LIST:
            self.read_head()
            n = self.read_int(0, False)
            self._read(n)
        else:
            raise ValueError("unknown tars type %s" % ttype)

    def skip_to_tag(self, tag, required=False):
        while self.remaining() > 0:
            ttype, cur = self.peek_head()
            if ttype == STRUCT_END:
                return False
            if cur == tag:
                return True
            if cur > tag:
                return False
            self.read_head()
            self.skip_field(ttype)
        if required:
            raise KeyError("missing tag %s" % tag)
        return False

    def read_bool(self, tag, required=False, default=False):
        return bool(self.read_int(tag, required, 1 if default else 0))

    def read_int(self, tag, required=False, default=0):
        if not self.skip_to_tag(tag, required):
            return default
        ttype, _ = self.read_head()
        if ttype == ZERO:
            return 0
        if ttype == BYTE:
            b = self._u8()
            return b - 256 if b > 127 else b
        if ttype == SHORT:
            return struct.unpack(">h", self._read(2))[0]
        if ttype == INT:
            return struct.unpack(">i", self._read(4))[0]
        if ttype == LONG:
            return struct.unpack(">q", self._read(8))[0]
        raise ValueError("not int type %s" % ttype)

    def read_string(self, tag, required=False, default=""):
        if not self.skip_to_tag(tag, required):
            return default
        ttype, _ = self.read_head()
        if ttype == ZERO:
            return ""
        if ttype == STRING1:
            n = self._u8()
            return self._read(n).decode("utf-8", "replace")
        if ttype == STRING4:
            n = struct.unpack(">I", self._read(4))[0]
            return self._read(n).decode("utf-8", "replace")
        raise ValueError("not string type %s" % ttype)

    def read_bytes(self, tag, required=False, default=b""):
        if not self.skip_to_tag(tag, required):
            return default
        ttype, _ = self.read_head()
        if ttype == ZERO:
            return b""
        if ttype == SIMPLE_LIST:
            self.read_head()
            n = self.read_int(0, False)
            return self._read(n)
        if ttype == LIST:
            n = self.read_int(0, False)
            out = bytearray()
            for _ in range(n):
                out.append(self.read_int(0, False) & 0xFF)
            return bytes(out)
        raise ValueError("not bytes type %s" % ttype)

    def read_struct_bytes(self, tag, required=False):
        if not self.skip_to_tag(tag, required):
            return None
        ttype, _ = self.read_head()
        if ttype != STRUCT_BEGIN:
            raise ValueError("not struct type %s" % ttype)
        start = self.pos
        depth = 1
        while depth > 0:
            nested, _ = self.read_head()
            if nested == STRUCT_BEGIN:
                depth += 1
            elif nested == STRUCT_END:
                depth -= 1
            else:
                self.skip_field(nested)
        return self.data[start:self.pos - 1]
